scipy/_lib/_array_api.py

```diff
@@ -191,33 +192,81 @@ def is_torch(xp):
     return xp.__name__ == 'scipy._lib.array_api_compat.array_api_compat.torch'
 
 
-def assert_equal(actual, desired, err_msg='', xp=None):
+def _strict_check(actual, desired, check_xp=True, check_dtype=True, check_shape=True):
+    try:
+        actual.dtype
+        desired.dtype
+    except AttributeError:
+        raise ValueError("Inputs should be arrays with a dtype attribute, "
+                         "python scalars and arrays are not accepted.")
```

Comment:
This is even more strict than I anticipated - perhaps uncomfortably strict even for me... but I think I can get over it. I'll reply again after thinking  bit. I'm guessing that I'll suggest for this check to be part of `check_xp`. If the user disables `check_xp`, at least `desired` can be automatically be made a NumPy type like `desired = np.asarray(desired)[()]`.

Reply:
Since the purpose of `check_xp` is to allow the developer to disable strict type checking, I would assume that with `check_xp=False`, a Python float would be OK. 

However, all the checks after this rely on `dtype` and `shape` attributes, so either they would fail, or the Python float would need to become a float (or 0d array) of the appropriate library.

Maybe if `check_xp=False`, then:
```python3
xp = array_namespace(actual)
desired = xp.asarray(desired)[()]
```
?

Reply:
- will `xp.asarray(desired)[()]` work for `torch` too?
- will it just be `xp.asarray(desired)` if we have a list instead of a scalar? If so, what function is used to check what we have?
- What about when `actual` is a 0-D array?

Reply:
> will `xp.asarray(desired)[()]` work for torch too?

It won't fail. This will always convert `desired` to an array except if `desired` is 0d and `xp` is NumPy. In that case, it will convert `desired` to a NumPy scalar.

> will it just be `xp.asarray(desired)` if we have a list instead of a scalar? If so, what function is used to check what we have?

`[()]` doesn't extract items out of an array with more than 0 dimensions, so I don't think you need to have a special case for lists vs scalars.

> What about when `actual` is a 0-D array?

I don't understand. `array_namespace(actual)` will still work even if `actual` is a 0d array, right? 

---

Perhaps all this is getting at what happens if `desired` is a 0D array? That is the only case I see in which this suggestion might be problematic because (for NumPy only) it would convert the 0d array to a scalar. If you don't want that, you could check whether `desired` is a NumPy array, and only then do the `[()]`. 

I neglected to include that step on purpose because IMO, SciPy functions should follow whatever NumPy's convention is, and vanilla NumPy's convention is to essentially *never* return a 0d array.
```python3
np.mean([1, 2, 3])  # scalar, not 0d array
np.asarray(0)*2  # scalar, not 0d array
np.sin(np.asarray(0))  # scalar, not 0d array
```
Strange as that may be, I think we should follow suit. So IF the developer were to pass a 0d NumPy array for `desired`, that would almost certainly be a mistake (since `actual` should never be a NumPy 0d array), and we are doing them the favor of converting the 0d array to a scalar. Then, this wil check to see whether `actual` is a scalar.

If NumPy's convention changes to eliminate scalars and start returning 0d arrays, then `xp.asarray(desired)[()]` will begin to return a 0d array, and that's fine, too. `desired` will become a 0d array, and we'll be checking whether `actual` is a 0d array. 

In summary, this will track the convention of `xp`: if `[()]` returns a 0d array, then we would want `actual` to be a 0d array instead of a scalar, and if `[()]` returns a scalar, we would want `actual` to be a scalar instead of a 0d array.

Reply:
If it helps, I guess `[()]` is not needed because this only happens *after* the `check_xp` block, and `[()]` would change neither the shape nor dtype.

Reply:
> assert_equal(0, np.asarray(0)[()])  # fails - good!

I'm unable to reproduce this failure from your example above... are you sure?

I don't know how to go about making this fail either.

Reply:
After https://github.com/scipy/scipy/pull/19186/commits/b99f0e716450eaae5134ceea8610d0d6e50a30b7, we now have improved behaviour with `check_xp=False`, e.g. `xp_assert_equal(np.asarray([1, 2]), torch.asarray([1, 2]), check_xp=False)` passes.

I have included `actual = xp.asarray(actual)` to enable dtype checks when `actual` is a built-in type.

However, I am struggling with built-in types for `check_xp=True`. `xp_assert_equal(np.asarray(0)[()], 0)`, `xp_assert_equal(0, np.asarray(0)[()])` and `xp_assert_equal(np.asarray([1, 2]), [1, 2])` all pass.

Also, the error message for `xp_assert_equal(1, [1, 2])` is `Desired an ndarray, but scalar given.`, which is not quite correct. We desire a list. This passes the rest of the function since `array_namespace` returns `array_api_compat.numpy` for all array-likes now.

Feel free to commit any changes that you need to get this working Matt!

Reply:
> I'm unable to reproduce this failure from your example above... are you sure?

No I can't reproduce that. Not sure what I was doing before...